// @ionic-native/speech-recognition/ngx/bundle.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var core$1 = require('@angular/core');
var core = require('@ionic-native/core');
require('rxjs');

var ɵngcc0 = require('@angular/core');
var SpeechRecognition = /** @class */ (function (_super) {
    tslib.__extends(SpeechRecognition, _super);
    function SpeechRecognition() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SpeechRecognition.prototype.isRecognitionAvailable = function () { return core.cordova(this, "isRecognitionAvailable", {}, arguments); };
    SpeechRecognition.prototype.startListening = function (options) { return core.cordova(this, "startListening", { "callbackOrder": "reverse", "observable": true }, arguments); };
    SpeechRecognition.prototype.stopListening = function () { return core.cordova(this, "stopListening", {}, arguments); };
    SpeechRecognition.prototype.getSupportedLanguages = function () { return core.cordova(this, "getSupportedLanguages", {}, arguments); };
    SpeechRecognition.prototype.hasPermission = function () { return core.cordova(this, "hasPermission", {}, arguments); };
    SpeechRecognition.prototype.requestPermission = function () { return core.cordova(this, "requestPermission", {}, arguments); };
    SpeechRecognition.pluginName = "SpeechRecognition";
    SpeechRecognition.plugin = "cordova-plugin-speechrecognition";
    SpeechRecognition.pluginRef = "plugins.speechRecognition";
    SpeechRecognition.repo = "https://github.com/pbakondy/cordova-plugin-speechrecognition";
    SpeechRecognition.platforms = ["Android", "iOS"];
SpeechRecognition.ɵfac = function SpeechRecognition_Factory(t) { return ɵSpeechRecognition_BaseFactory(t || SpeechRecognition); };
SpeechRecognition.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: SpeechRecognition, factory: function (t) { return SpeechRecognition.ɵfac(t); } });
var ɵSpeechRecognition_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(SpeechRecognition);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SpeechRecognition, [{
        type: core$1.Injectable
    }], null, null); })();
    return SpeechRecognition;
}(core.IonicNativePlugin));

exports.SpeechRecognition = SpeechRecognition;
