// angular-bottom-sheet/esm5/app/angular-bottom-sheet.module.js
import * as tslib_1 from "tslib";
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { AngularBottomSheetComponent } from './component/angular-bottom-sheet.component';
import { TestingComponent } from './testing/testing.component';
import * as ɵngcc0 from '@angular/core';
var AngularBottomSheetModule = /** @class */ (function () {
    function AngularBottomSheetModule() {
    }
AngularBottomSheetModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularBottomSheetModule, bootstrap: function () { return [TestingComponent]; } });
AngularBottomSheetModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularBottomSheetModule_Factory(t) { return new (t || AngularBottomSheetModule)(); }, providers: [], imports: [[
            BrowserModule,
            BrowserAnimationsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularBottomSheetModule, { declarations: function () { return [AngularBottomSheetComponent,
        TestingComponent]; }, imports: function () { return [BrowserModule,
        BrowserAnimationsModule]; }, exports: function () { return [AngularBottomSheetComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularBottomSheetModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    AngularBottomSheetComponent,
                    TestingComponent,
                ],
                imports: [
                    BrowserModule,
                    BrowserAnimationsModule
                ],
                exports: [
                    AngularBottomSheetComponent
                ],
                providers: [],
                bootstrap: [TestingComponent]
            }]
    }], function () { return []; }, null); })();
    return AngularBottomSheetModule;
}());
export { AngularBottomSheetModule };
