// angular-bottom-sheet/esm5/app/testing/testing.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { AngularBottomSheetComponent } from '../component/angular-bottom-sheet.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../component/angular-bottom-sheet.component';

var _c0 = ["test"];
var TestingComponent = /** @class */ (function () {
    function TestingComponent() {
    }
    TestingComponent.prototype.ngOnInit = function () {
        this.options = {
            title: "Angular Bottom Sheet",
            enableCloseButton: true,
            darkTheme: false
        };
        this.open();
    };
    TestingComponent.prototype.toggle = function () {
        this.testing.toggle();
    };
    TestingComponent.prototype.open = function () {
        this.testing.open();
    };
    TestingComponent.prototype.close = function () {
        this.testing.close();
    };
    TestingComponent.prototype.toggleDarkTheme = function () {
        this.open();
        this.testing.toggleDarkTheme();
    };
    TestingComponent.prototype.toggleCloseButton = function () {
        this.open();
        this.testing.toggleCloseButton();
    };
    tslib_1.__decorate([
        ViewChild('test'),
        tslib_1.__metadata("design:type", AngularBottomSheetComponent)
    ], TestingComponent.prototype, "testing", void 0);
    TestingComponent = tslib_1.__decorate([ tslib_1.__metadata("design:paramtypes", [])
    ], TestingComponent);
TestingComponent.ɵfac = function TestingComponent_Factory(t) { return new (t || TestingComponent)(); };
TestingComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: TestingComponent, selectors: [["app-root"]], viewQuery: function TestingComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.testing = _t.first);
    } }, decls: 15, vars: 1, consts: [[1, "button-container"], ["type", "button", 1, "btn", "btn-outline-primary", "btn-sm", 3, "click"], ["type", "button", 1, "btn", "btn-outline-danger", "btn-sm", 3, "click"], ["type", "button", 1, "btn", "btn-outline-dark", "btn-sm", 3, "click"], [3, "options"], ["test", ""]], template: function TestingComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_1_listener() { return ctx.open(); });
        ɵngcc0.ɵɵtext(2, "Open");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "button", 2);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_3_listener() { return ctx.close(); });
        ɵngcc0.ɵɵtext(4, "Close");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_5_listener() { return ctx.toggle(); });
        ɵngcc0.ɵɵtext(6, "Toggle");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "button", 3);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_7_listener() { return ctx.toggleDarkTheme(); });
        ɵngcc0.ɵɵtext(8, "Toggle Dark Theme");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_9_listener() { return ctx.toggleCloseButton(); });
        ɵngcc0.ɵɵtext(10, "Toggle Close Button");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "angular-bottom-sheet", 4, 5);
        ɵngcc0.ɵɵelementStart(13, "p");
        ɵngcc0.ɵɵtext(14, "What is Lorem Ipsum?");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(11);
        ɵngcc0.ɵɵproperty("options", ctx.options);
    } }, directives: [ɵngcc1.AngularBottomSheetComponent], styles: [".button-container[_ngcontent-%COMP%]{z-index:9999;position:absolute;background:#fff;right:0;left:0;text-align:center}.button-container[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]{margin:8px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TestingComponent, [{
        type: Component,
        args: [{
                selector: 'app-root',
                template: "<div class=\"button-container\">\n  <button type=\"button\" class=\"btn btn-outline-primary  btn-sm\" (click)=\"open()\">Open</button>\n  <button type=\"button\" class=\"btn btn-outline-danger btn-sm\" (click)=\"close()\">Close</button>\n  <button type=\"button\" class=\"btn btn-outline-primary btn-sm\" (click)=\"toggle()\">Toggle</button>\n  <button type=\"button\" class=\"btn btn-outline-dark btn-sm\" (click)=\"toggleDarkTheme()\">Toggle Dark Theme</button>\n  <button type=\"button\" class=\"btn btn-outline-primary btn-sm\" (click)=\"toggleCloseButton()\">Toggle Close Button</button>\n</div>\n\n<angular-bottom-sheet #test [options]=\"options\">\n  <p>What is Lorem Ipsum?</p>\n</angular-bottom-sheet>\n",
                styles: [".button-container{z-index:9999;position:absolute;background:#fff;right:0;left:0;text-align:center}.button-container button{margin:8px}"]
            }]
    }], function () { return []; }, { testing: [{
            type: ViewChild,
            args: ['test']
        }] }); })();
    return TestingComponent;
}());
export { TestingComponent };
