// angular-bottom-sheet/bundles/angular-bottom-sheet.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/platform-browser'),require('@angular/platform-browser/animations'),exports, require('@angular/platform-browser'), require('@angular/core'), require('@angular/platform-browser/animations'), require('@angular/animations')) :
    typeof define === 'function' && define.amd ? define('angular-bottom-sheet', ['@angular/core','@angular/common','@angular/platform-browser','@angular/platform-browser/animations','exports', '@angular/platform-browser', '@angular/core', '@angular/platform-browser/animations', '@angular/animations'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,global.ng.platformBrowser,global.ng.platformBrowser.animations,global['angular-bottom-sheet'] = {}, global.ng.platformBrowser, global.ng.core, global.ng.platformBrowser.animations, global.ng.animations));
}(this, function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, platformBrowser, core, animations, animations$1) { 
function AngularBottomSheetComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    var _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵlistener("click", function AngularBottomSheetComponent_div_0_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.close(); });
    ɵngcc0.ɵɵelementEnd();
} }
var _c0 = function (a0) { return { "bg-dark-footer": a0 }; };
function AngularBottomSheetComponent_div_1_div_7_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵlistener("click", function AngularBottomSheetComponent_div_1_div_7_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); var ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.close(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c0, ctx_r4.options == null ? null : ctx_r4.options.darkTheme));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r4.options.closeButtonTitle, " ");
} }
var _c1 = function (a0) { return { "bg-dark": a0 }; };
var _c2 = function (a0) { return { "bg-dark-header": a0 }; };
function AngularBottomSheetComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementStart(2, "div", 4);
    ɵngcc0.ɵɵelementStart(3, "h4", 5);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 6);
    ɵngcc0.ɵɵprojection(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, AngularBottomSheetComponent_div_1_div_7_Template, 2, 4, "div", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("background-color", ctx_r1.options == null ? null : ctx_r1.options.backgroundColor)("color", ctx_r1.options == null ? null : ctx_r1.options.fontColor);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c1, ctx_r1.options == null ? null : ctx_r1.options.darkTheme))("@slideUpToggle", undefined);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(11, _c2, ctx_r1.options == null ? null : ctx_r1.options.darkTheme));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.options == null ? null : ctx_r1.options.title);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.flags.isCloseButtonEnabled);
} }
var _c3 = ["*"];
var _c4 = ["test"];
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    var SlideUpToggleAnimation = [
        animations$1.trigger('slideUpToggle', [
            animations$1.transition(':enter', [
                animations$1.style({ transform: 'translateY(100%)' }),
                animations$1.animate('200ms ease-in', animations$1.style({ transform: 'translateY(0%)' }))
            ]),
            animations$1.transition(':leave', [
                animations$1.animate('200ms ease-in', animations$1.style({ transform: 'translateY(100%)' }))
            ])
        ])
    ];

    /**
     * Component
     */
    var AngularBottomSheetComponent = /** @class */ (function () {
        function AngularBottomSheetComponent(changeDetector) {
            this.changeDetector = changeDetector;
            this.flags = {
                isBottomSheetEnabled: false
            };
        }
        AngularBottomSheetComponent.prototype.ngOnInit = function () {
            this.flags.isCloseButtonEnabled = this.options.enableCloseButton ? true : false;
            this.options.closeButtonTitle = this.options.closeButtonTitle ? this.options.closeButtonTitle : "Close";
        };
        /**
         * Opens bottom sheet component
         */
        AngularBottomSheetComponent.prototype.open = function () {
            this.flags.isBottomSheetEnabled = true;
            this.changeDetector.detectChanges();
        };
        /**
         * Closes bottom sheet component
         */
        AngularBottomSheetComponent.prototype.close = function () {
            this.flags.isBottomSheetEnabled = false;
            this.changeDetector.detectChanges();
        };
        /**
         * Toggles bottom sheet component
         */
        AngularBottomSheetComponent.prototype.toggle = function () {
            this.flags.isBottomSheetEnabled = !this.flags.isBottomSheetEnabled;
            this.changeDetector.detectChanges();
        };
        /**
         * Toggles close button
         */
        AngularBottomSheetComponent.prototype.toggleCloseButton = function () {
            this.flags.isCloseButtonEnabled = !this.flags.isCloseButtonEnabled;
            this.changeDetector.detectChanges();
        };
        /**
         * Toggles dark theme
         */
        AngularBottomSheetComponent.prototype.toggleDarkTheme = function () {
            this.options.darkTheme = !this.options.darkTheme;
            this.changeDetector.detectChanges();
        };
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], AngularBottomSheetComponent.prototype, "options", void 0);
        AngularBottomSheetComponent = __decorate([ __metadata("design:paramtypes", [core.ChangeDetectorRef])
        ], AngularBottomSheetComponent);
AngularBottomSheetComponent.ɵfac = function AngularBottomSheetComponent_Factory(t) { return new (t || AngularBottomSheetComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
AngularBottomSheetComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AngularBottomSheetComponent, selectors: [["angular-bottom-sheet"]], inputs: { options: "options" }, ngContentSelectors: _c3, decls: 2, vars: 2, consts: [["class", "bottom-sheet-bg-overlay", 3, "click", 4, "ngIf"], ["class", "bottom-sheet-container", 3, "ngClass", "backgroundColor", "color", 4, "ngIf"], [1, "bottom-sheet-bg-overlay", 3, "click"], [1, "bottom-sheet-container", 3, "ngClass"], [1, "bottom-sheet-header", 3, "ngClass"], [1, "title"], [1, "bottom-sheet-content"], ["class", "bottom-sheet-footer", 3, "ngClass", "click", 4, "ngIf"], [1, "bottom-sheet-footer", 3, "ngClass", "click"]], template: function AngularBottomSheetComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, AngularBottomSheetComponent_div_0_Template, 1, 0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, AngularBottomSheetComponent_div_1_Template, 8, 13, "div", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.flags.isBottomSheetEnabled);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.flags.isBottomSheetEnabled);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass], styles: [".bottom-sheet-bg-overlay[_ngcontent-%COMP%]{background:rgba(0,0,0,.32);position:absolute;top:0;bottom:0;left:0;right:0}.bottom-sheet-container[_ngcontent-%COMP%]{background:#fff;position:fixed;bottom:0;border-radius:8px 8px 0 0;left:50%;margin-left:-35%;width:75%;z-index:4;color:#363636;box-shadow:0 8px 10px -5px rgba(0,0,0,.2),0 16px 24px 2px rgba(0,0,0,.14),0 6px 30px 5px rgba(0,0,0,.12);font-family:Roboto,sans-serif}.bg-dark[_ngcontent-%COMP%]{color:rgba(255,255,255,.9);background:#363636}.bottom-sheet-header[_ngcontent-%COMP%]   h4.title[_ngcontent-%COMP%]{display:inline-block;font-weight:500;margin:0;font-size:1rem}.bottom-sheet-header[_ngcontent-%COMP%]{border-bottom:1px solid rgba(0,0,0,.16);padding:16px}.bg-dark-header[_ngcontent-%COMP%]{border-bottom:1px solid rgba(255,255,255,.16)}.bottom-sheet-footer[_ngcontent-%COMP%]{text-align:center;padding:16px;border-top:1px solid rgba(0,0,0,.16);background:rgba(0,0,0,.02);cursor:pointer}.bg-dark-footer[_ngcontent-%COMP%]{border-top:1px solid rgba(255,255,255,.16)}.bottom-sheet-content[_ngcontent-%COMP%]{max-height:300px;overflow-y:auto;padding:16px}[_ngcontent-%COMP%]::-webkit-scrollbar{width:4px}[_ngcontent-%COMP%]::-webkit-scrollbar-track{background:#bdbdbd}[_ngcontent-%COMP%]::-webkit-scrollbar-thumb{background:#888}[_ngcontent-%COMP%]::-webkit-scrollbar-thumb:hover{background:#5c5c5c}@media (max-width:992px){.bottom-sheet-container[_ngcontent-%COMP%]{width:100%;position:fixed;bottom:0;left:0;right:0;margin-left:0}}"], data: { animation: [SlideUpToggleAnimation] }, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularBottomSheetComponent, [{
        type: core.Component,
        args: [{
                selector: 'angular-bottom-sheet',
                template: "<div class=\"bottom-sheet-bg-overlay\" *ngIf=\"flags.isBottomSheetEnabled\" (click)=\"close()\"></div>\r\n<div class=\"bottom-sheet-container\" [ngClass]=\"{'bg-dark': options?.darkTheme}\" *ngIf=\"flags.isBottomSheetEnabled\"\r\n  [style.backgroundColor]=\"options?.backgroundColor\" [style.color]=\"options?.fontColor\" [@slideUpToggle]>\r\n  <div>\r\n    <div class=\"bottom-sheet-header\" [ngClass]=\"{'bg-dark-header': options?.darkTheme}\">\r\n      <h4 class=\"title\">{{options?.title}}</h4>\r\n    </div>\r\n    <div class=\"bottom-sheet-content\">\r\n      <ng-content></ng-content>\r\n    </div>\r\n    <div class=\"bottom-sheet-footer\" [ngClass]=\"{'bg-dark-footer': options?.darkTheme}\" (click)=\"close()\" *ngIf=\"flags.isCloseButtonEnabled\">\r\n      {{options.closeButtonTitle}}\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                animations: [SlideUpToggleAnimation],
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [".bottom-sheet-bg-overlay{background:rgba(0,0,0,.32);position:absolute;top:0;bottom:0;left:0;right:0}.bottom-sheet-container{background:#fff;position:fixed;bottom:0;border-radius:8px 8px 0 0;left:50%;margin-left:-35%;width:75%;z-index:4;color:#363636;box-shadow:0 8px 10px -5px rgba(0,0,0,.2),0 16px 24px 2px rgba(0,0,0,.14),0 6px 30px 5px rgba(0,0,0,.12);font-family:Roboto,sans-serif}.bg-dark{color:rgba(255,255,255,.9);background:#363636}.bottom-sheet-header h4.title{display:inline-block;font-weight:500;margin:0;font-size:1rem}.bottom-sheet-header{border-bottom:1px solid rgba(0,0,0,.16);padding:16px}.bg-dark-header{border-bottom:1px solid rgba(255,255,255,.16)}.bottom-sheet-footer{text-align:center;padding:16px;border-top:1px solid rgba(0,0,0,.16);background:rgba(0,0,0,.02);cursor:pointer}.bg-dark-footer{border-top:1px solid rgba(255,255,255,.16)}.bottom-sheet-content{max-height:300px;overflow-y:auto;padding:16px}::-webkit-scrollbar{width:4px}::-webkit-scrollbar-track{background:#bdbdbd}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb:hover{background:#5c5c5c}@media (max-width:992px){.bottom-sheet-container{width:100%;position:fixed;bottom:0;left:0;right:0;margin-left:0}}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { options: [{
            type: core.Input
        }] }); })();
        return AngularBottomSheetComponent;
    }());

    var TestingComponent = /** @class */ (function () {
        function TestingComponent() {
        }
        TestingComponent.prototype.ngOnInit = function () {
            this.options = {
                title: "Angular Bottom Sheet",
                enableCloseButton: true,
                darkTheme: false
            };
            this.open();
        };
        TestingComponent.prototype.toggle = function () {
            this.testing.toggle();
        };
        TestingComponent.prototype.open = function () {
            this.testing.open();
        };
        TestingComponent.prototype.close = function () {
            this.testing.close();
        };
        TestingComponent.prototype.toggleDarkTheme = function () {
            this.open();
            this.testing.toggleDarkTheme();
        };
        TestingComponent.prototype.toggleCloseButton = function () {
            this.open();
            this.testing.toggleCloseButton();
        };
        __decorate([
            core.ViewChild('test'),
            __metadata("design:type", AngularBottomSheetComponent)
        ], TestingComponent.prototype, "testing", void 0);
        TestingComponent = __decorate([ __metadata("design:paramtypes", [])
        ], TestingComponent);
TestingComponent.ɵfac = function TestingComponent_Factory(t) { return new (t || TestingComponent)(); };
TestingComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: TestingComponent, selectors: [["app-root"]], viewQuery: function TestingComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c4, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.testing = _t.first);
    } }, decls: 15, vars: 1, consts: [[1, "button-container"], ["type", "button", 1, "btn", "btn-outline-primary", "btn-sm", 3, "click"], ["type", "button", 1, "btn", "btn-outline-danger", "btn-sm", 3, "click"], ["type", "button", 1, "btn", "btn-outline-dark", "btn-sm", 3, "click"], [3, "options"], ["test", ""]], template: function TestingComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_1_listener() { return ctx.open(); });
        ɵngcc0.ɵɵtext(2, "Open");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "button", 2);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_3_listener() { return ctx.close(); });
        ɵngcc0.ɵɵtext(4, "Close");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_5_listener() { return ctx.toggle(); });
        ɵngcc0.ɵɵtext(6, "Toggle");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "button", 3);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_7_listener() { return ctx.toggleDarkTheme(); });
        ɵngcc0.ɵɵtext(8, "Toggle Dark Theme");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TestingComponent_Template_button_click_9_listener() { return ctx.toggleCloseButton(); });
        ɵngcc0.ɵɵtext(10, "Toggle Close Button");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "angular-bottom-sheet", 4, 5);
        ɵngcc0.ɵɵelementStart(13, "p");
        ɵngcc0.ɵɵtext(14, "What is Lorem Ipsum?");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(11);
        ɵngcc0.ɵɵproperty("options", ctx.options);
    } }, directives: [AngularBottomSheetComponent], styles: [".button-container[_ngcontent-%COMP%]{z-index:9999;position:absolute;background:#fff;right:0;left:0;text-align:center}.button-container[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]{margin:8px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TestingComponent, [{
        type: core.Component,
        args: [{
                selector: 'app-root',
                template: "<div class=\"button-container\">\n  <button type=\"button\" class=\"btn btn-outline-primary  btn-sm\" (click)=\"open()\">Open</button>\n  <button type=\"button\" class=\"btn btn-outline-danger btn-sm\" (click)=\"close()\">Close</button>\n  <button type=\"button\" class=\"btn btn-outline-primary btn-sm\" (click)=\"toggle()\">Toggle</button>\n  <button type=\"button\" class=\"btn btn-outline-dark btn-sm\" (click)=\"toggleDarkTheme()\">Toggle Dark Theme</button>\n  <button type=\"button\" class=\"btn btn-outline-primary btn-sm\" (click)=\"toggleCloseButton()\">Toggle Close Button</button>\n</div>\n\n<angular-bottom-sheet #test [options]=\"options\">\n  <p>What is Lorem Ipsum?</p>\n</angular-bottom-sheet>\n",
                styles: [".button-container{z-index:9999;position:absolute;background:#fff;right:0;left:0;text-align:center}.button-container button{margin:8px}"]
            }]
    }], function () { return []; }, { testing: [{
            type: core.ViewChild,
            args: ['test']
        }] }); })();
        return TestingComponent;
    }());

    var AngularBottomSheetModule = /** @class */ (function () {
        function AngularBottomSheetModule() {
        }
AngularBottomSheetModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularBottomSheetModule, bootstrap: [TestingComponent] });
AngularBottomSheetModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularBottomSheetModule_Factory(t) { return new (t || AngularBottomSheetModule)(); }, providers: [], imports: [[
            platformBrowser.BrowserModule,
            animations.BrowserAnimationsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularBottomSheetModule, { declarations: [AngularBottomSheetComponent, TestingComponent], imports: [ɵngcc2.BrowserModule, ɵngcc3.BrowserAnimationsModule], exports: [AngularBottomSheetComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularBottomSheetModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    AngularBottomSheetComponent,
                    TestingComponent,
                ],
                imports: [
                    platformBrowser.BrowserModule,
                    animations.BrowserAnimationsModule
                ],
                exports: [
                    AngularBottomSheetComponent
                ],
                providers: [],
                bootstrap: [TestingComponent]
            }]
    }], function () { return []; }, null); })();
        return AngularBottomSheetModule;
    }());

    exports.AngularBottomSheetComponent = AngularBottomSheetComponent;
    exports.AngularBottomSheetModule = AngularBottomSheetModule;
    exports.ɵa = SlideUpToggleAnimation;
    exports.ɵb = TestingComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
